### ISSUE TITLE
cutlass's gemm produces different result from Pytorch's gemm

### ISSUE THREAD

**Issue by reveriel** [NVIDIA/cutlass#282]
I compared cutlass's GEMM(see below, copied from `cutlass/test/unit/gemm/device/simt_sgemm_tt_sm50.cu`)
with pytorch's matrix multiplication( torch.mm ), got slightly different results. 

I found:
- The difference is deterministic.
- The difference is relevant to the f32 representation. Computing numbers that can be represented exactly using f32 format has zero difference.

Such difference has a significant impact on my application.
Why does such difference exist?  Did I do something wrong and how can I get the exact same result?

``` c++
    using Gemm = cutlass::gemm::device::Gemm<
        precision, cutlass::layout::RowMajor,
        precision, cutlass::layout::RowMajor,
        precision, cutlass::layout::RowMajor,
        precision,
        cutlass::arch::OpClassSimt,
        cutlass::arch::Sm50,
        ThreadblockShape, WarpShape, InstructionShape,
        EpilogueOutputOp,
        cutlass::gemm::threadblock::GemmIdentityThreadblockSwizzle<>,
        2 // Stages
        >;
```

I'm using the lateset cutlass (commit `856d4db`). PyTorch 1.5, CUDA 10.2.

here is the code.
``` c++
#include "cutlass/cutlass.h"
#include "cutlass/gemm/device/gemm.h"
#include "cutlass/gemm/gemm.h"
#include "cutlass/numeric_types.h"

#include "cutlass/util/host_tensor.h"

#include <cuda.h>
#include <cuda_runtime.h>
#include <torch/extension.h>


inline char const* to_string(cutlass::Status status)
{

    switch (status) {
    case cutlass::Status::kSuccess:
        return "kSuccess";
    case cutlass::Status::kErrorMisalignedOperand:
        return "kErrorMisalignedOperand";
    case cutlass::Status::kErrorInvalidLayout:
        return "kErrorInvalidLayout";
    case cutlass::Status::kErrorInvalidProblem:
        return "kErrorInvalidProblem";
    case cutlass::Status::kErrorNotSupported:
        return "kErrorNotSupported";
    case cutlass::Status::kErrorWorkspaceNull:
        return "kErrorWorkspaceNull";
    case cutlass::Status::kErrorInternal:
        return "kErrorInternal";
    case cutlass::Status::kInvalid:
        return "kInvalid";
    default:
        break;
    }
    return "invalid";
}

template <typename Gemm>
class Test
{
    using ElementCompute = typename Gemm::EpilogueOutputOp::ElementCompute;

public:
    typename Gemm::TensorRefA ref_A;
    typename Gemm::TensorRefB ref_B;
    typename Gemm::TensorRefC ref_C;
    typename Gemm::TensorRefD ref_D;

    Test(torch::Tensor A,
         torch::Tensor B,
         torch::Tensor C)
    {
        auto a = A.packed_accessor64<float, 2, torch::RestrictPtrTraits>();
        auto b = B.packed_accessor64<float, 2, torch::RestrictPtrTraits>();
        auto c = C.packed_accessor64<float, 2, torch::RestrictPtrTraits>();

        ref_A = typename Gemm::TensorRefA((float*)A.data_ptr(), cutlass::layout::RowMajor(a.stride(0)));
        ref_B = typename Gemm::TensorRefB((float*)B.data_ptr(), cutlass::layout::RowMajor(b.stride(0)));
        ref_C = typename Gemm::TensorRefC((float*)C.data_ptr(), cutlass::layout::RowMajor(c.stride(0)));
        ref_D = typename Gemm::TensorRefD((float*)C.data_ptr(), cutlass::layout::RowMajor(c.stride(0)));
    }

    bool run(
        cutlass::gemm::GemmCoord problem_size,
        int split_k_slices = 1,
        ElementCompute alpha = ElementCompute(1),
        ElementCompute beta = ElementCompute(0))
    {

        // D = A * B + C
        typename Gemm::Arguments arguments{
            problem_size,
            ref_A,
            ref_B,
            ref_C,
            ref_D,
            {alpha, beta},
            split_k_slices};

        Gemm gemm_op;

        size_t workspace_size = Gemm::get_workspace_size(arguments);
        cutlass::device_memory::allocation<uint8_t> workspace(workspace_size);
        cutlass::Status status = gemm_op.initialize(arguments, workspace.get());

        if (status != cutlass::Status::kSuccess) {
            cudaError_t error = cudaGetLastError();
            std::cerr << "This test is not supported: " << cudaGetErrorString(error) << "\n";
            return true;
        }

        //
        // Run the GEMM
        //

        status = gemm_op();

        std::cout << to_string(status) << std::endl;
        // EXPECT_TRUE(status == cutlass::Status::kSuccess) << to_string(status);
        return status == cutlass::Status::kSuccess;
    }
};

template <int M, int N, int K>
void test_cutlass(
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor C)
{
    using precision = float;
    using ThreadblockShape = cutlass::gemm::GemmShape<8, 32, 8>;
    using WarpShape = cutlass::gemm::GemmShape<8, 32, 8>;

    static int const kEpilogueElementsPerAccess = 1;
    using InstructionShape = cutlass::gemm::GemmShape<1, 1, 1>;
    using EpilogueOutputOp = cutlass::epilogue::thread::LinearCombination<
        precision, kEpilogueElementsPerAccess, precision, precision>;

    using Gemm = cutlass::gemm::device::Gemm<
        precision, cutlass::layout::RowMajor,
        precision, cutlass::layout::RowMajor,
        precision, cutlass::layout::RowMajor,
        precision,
        cutlass::arch::OpClassSimt,
        cutlass::arch::Sm50,
        ThreadblockShape, WarpShape, InstructionShape,
        EpilogueOutputOp,
        cutlass::gemm::threadblock::GemmIdentityThreadblockSwizzle<>,
        2 // Stages
        >;

    cutlass::gemm::GemmCoord problem_size(M, N, K);

    Test<Gemm> test(A, B, C);
    double alpha = 1;
    double beta = 1;
    int split_k = 1;

    using ElementCompute = typename Gemm::EpilogueOutputOp::ElementCompute;

    //  D = alpha * A * B + beta * C
    test.run(problem_size, split_k,
             cutlass::from_real<ElementCompute>(alpha),
             cutlass::from_real<ElementCompute>(beta));
}

int main()
{
    const int M = 8;
    const int N = 32;
    const int K = 64;

    torch::manual_seed(0);
    auto options = torch::TensorOptions().dtype(torch::kFloat32).layout(torch::kStrided).device(torch::kCUDA, 0).requires_grad(false);
    // torch::Tensor A = torch::randint(40, {M, K}, options);
    // torch::Tensor B = torch::randint(40, {K, N}, options);
    torch::Tensor A = torch::randn({M, K}, options);
    torch::Tensor B = torch::randn({K, N}, options);
    torch::Tensor C_torch = torch::zeros({M, N}, options);
    torch::mm_out(C_torch, A, B);

    torch::Tensor C_cutlass = torch::zeros({M, N}, options);
    test_cutlass<M, N, K>(A, B, C_cutlass);


    std::cout << " C_torch = \n" << C_torch << std::endl;
    std::cout << " cutlass result C2 = \n" << C_cutlass << std::endl;

    std::cout << "distance = " << (C_cutlass - C_torch).abs().sum() << std::endl;

    return 0;
}
```

running this produces 
```
distance = 0.000205257
```

**Comment by hwu36** [comment]
I don't know what implementation is used by pytorch.  Generally, float point does not support associative.  For example, `f1 + f2 + f3 + f4` is different from `(f1 + f2) + (f3 + f4)` no matter running on CPU or GPU.  Also the internal accumulation type used by different hardware and can also be different. So, we cannot compare two fp32 gemm results bit by bit.  Instead, we need to allow them to have some epsilon. 

If the precision is important to you, you can use fp64.

**Comment by reveriel** [comment]
pytorch uses cublas in this case。

thank you. I gotit.

**Comment by hwu36** [comment]
you can use nvprof/nsight to check what kernel is used by cublas/pytorch.

**Comment by reveriel** [comment]
pytorch uses `gemmSN_NN_kernel`

Is there any standard describing how accurate a gemm implementation is?

**Comment by cliffburdick** [comment]
@reveriel take a look at this: https://docs.nvidia.com/cuda/floating-point/index.html

I'm not aware of any standard. It's certainly possible there is a bug where an answer could be more accurate if the order of operations was changed, but you would probably need to provide more info. Have you tried double precision and compared it to single precision to see how far off each one is?

**Comment by timespacce** [comment]
One of the biggest problems of modern hardware is the absence of absolute precision.

Take a look at the simple 32bit float expression - **0.2 + 0.1** - it returns on any modern hardware **0.30000000000000004**. Now imagine thousands of numbers being multiplied and summed. The error gets bigger and bigger after each multiplication/addition.

So the question is - which implementation returns more accurate results? Take some high-accuracy-library and compute the result and then compore to both cutlass and pytorch. I wouldn't exclude the chance of pytorch having worse precision than cutlass. Tensorflow for example is full of non-deterministic or less accurate primitives.

**Comment by hwu36** [comment]
Thank you, everyone.

Two more question about precision:

1. Do you know any important application that fp32 is not accurate enough, but fp64 is overkill?
2. Do you know any important fp32 application that tf32 introduced by Ampere is not accurate enough?

**Comment by reveriel** [comment]
I use modified a sparse convolution implementation by replacing the torch.mm with cutlass's gemm (and some other changes). 

These two versions process slightly different results using the same filter weight and input.  (no difference if weight and input are losslessly representable by fp32). 

But replacing the sparse convolution in a neural network results in a large drop in the final mAP. (I didn't trained a new weight.) I guess this is caused by accumulated errors of this epsilon.

fp64 is not a viable solution because I'm using a RTX2080Ti. FP64 : FP32 throughput is 1:32.

I think I need to implement the backprop for the sparse convolution and train a new weight.

**Comment by reveriel** [comment]
I compared cutlass's fp32 gemm with pytorch's(cublas) fp32 gemm, using pytorch's fp64 as reference.

Seems pytorch is more accurate 

``` c++
cutlass distance = 0.0215418
torch distance = 0.0142782
```

code:
```c++
int main()
{
    const int M = 8;
    const int N = 32;
    const int K = 64;

    torch::manual_seed(0);
    double cutlass_distance = 0;
    double torch_distance = 0;

    auto options_64 = torch::TensorOptions().dtype(torch::kFloat64).layout(torch::kStrided).device(torch::kCUDA, 0).requires_grad(false);
    auto options_32 = torch::TensorOptions().dtype(torch::kFloat32).layout(torch::kStrided).device(torch::kCUDA, 0).requires_grad(false);

    for (int i = 0; i < 100; i++) {
        torch::Tensor A_64 = torch::randn({M, K}, options_64);
        torch::Tensor B_64 = torch::randn({K, N}, options_64);
        torch::Tensor C_64 = torch::zeros({M, N}, options_64);
        torch::mm_out(C_64, A_64, B_64);

        torch::Tensor A = A_64.to(torch::kFloat32);
        torch::Tensor B = B_64.to(torch::kFloat32);
        torch::Tensor C_torch = torch::zeros({M, N}, options_32);
        torch::mm_out(C_torch, A, B);

        torch::Tensor C_cutlass = torch::zeros({M, N}, options_32);
        test_cutlass<M, N, K>(A, B, C_cutlass);

        cutlass_distance += (C_cutlass.to(torch::kFloat64) - C_64).abs().sum().item<double>();
        torch_distance += (C_torch.to(torch::kFloat64) - C_64).abs().sum().item<double>();
    }

    std::cout << "cutlass distance = " << cutlass_distance << std::endl;
    std::cout << "torch distance = " << torch_distance << std::endl;

    return 0;
}
```

**Comment by cliffburdick** [comment]
> I compared cutlass's fp32 gemm with pytorch's(cublas) fp32 gemm, using pytorch's fp64 as reference.
> 
> Seems pytorch is more accurate
> 
> ```
> cutlass distance = 0.0215418
> torch distance = 0.0142782
> ```
> 
> It's interesting that the ratio of them is always around 3:2
> 
> removing the `abs()`, I got this:
> 
> ```
> cutlass distance = 0.000325772
> torch distance = 6.48636e-05
> ```
> 
> code:
> 
> ```c++
> int main()
> {
>     const int M = 8;
>     const int N = 32;
>     const int K = 64;
> 
>     torch::manual_seed(0);
>     double cutlass_distance = 0;
>     double torch_distance = 0;
> 
>     auto options_64 = torch::TensorOptions().dtype(torch::kFloat64).layout(torch::kStrided).device(torch::kCUDA, 0).requires_grad(false);
>     auto options_32 = torch::TensorOptions().dtype(torch::kFloat32).layout(torch::kStrided).device(torch::kCUDA, 0).requires_grad(false);
> 
>     for (int i = 0; i < 100; i++) {
>         torch::Tensor A_64 = torch::randn({M, K}, options_64);
>         torch::Tensor B_64 = torch::randn({K, N}, options_64);
>         torch::Tensor C_64 = torch::zeros({M, N}, options_64);
>         torch::mm_out(C_64, A_64, B_64);
> 
>         torch::Tensor A = A_64.to(torch::kFloat32);
>         torch::Tensor B = B_64.to(torch::kFloat32);
>         torch::Tensor C_torch = torch::zeros({M, N}, options_32);
>         torch::mm_out(C_torch, A, B);
> 
>         torch::Tensor C_cutlass = torch::zeros({M, N}, options_32);
>         test_cutlass<M, N, K>(A, B, C_cutlass);
> 
>         cutlass_distance += (C_cutlass.to(torch::kFloat64) - C_64).abs().sum().item<double>();
>         torch_distance += (C_torch.to(torch::kFloat64) - C_64).abs().sum().item<double>();
>     }
> 
>     std::cout << "cutlass distance = " << cutlass_distance << std::endl;
>     std::cout << "torch distance = " << torch_distance << std::endl;
> 
>     return 0;
> }
> ```
@reveriel what is the magnitude of these floating point numbers?

**Comment by reveriel** [comment]
> @reveriel what is the magnitude of these floating point numbers?
@cliffburdick  `torch::randn` fills tensors with random numbers from a normal distribution with mean 0 and variance 1
